vertx-sql-client/src/main/java/io/vertx/sqlclient/PoolConfig.java

```diff
@@ -0,0 +1,129 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.vertx.sqlclient;
+
+import io.vertx.codegen.annotations.Fluent;
+import io.vertx.codegen.annotations.GenIgnore;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.Handler;
+import io.vertx.sqlclient.impl.PoolConfigImpl;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * The pool configuration that comprehends several mandatory configuration items, plus a few extra optional ones.
+ *
+ * The pool config provides advanced features:
+ *
+ * <ul>
+ *   <li>Connect handler for connection initialization</li>
+ *   <li>Load balance across multiple servers</li>
+ * </ul>
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+@VertxGen
+public interface PoolConfig {
+
+  /**
+   * Create a pool config.
+   *
+   * @param options the pool options
+   */
+  static PoolConfig create(PoolOptions options) {
+    return new PoolConfigImpl(options);
+  }
+
+  /**
+   * Create a pool config.
+   */
+  static PoolConfig create() {
+    return create(new PoolOptions());
+  }
+
+  /**
+   * Configures the pool to connect to given {@code server}.
+   *
+   * @param server the connect options
+   * @return a reference to this, so the API can be used fluently
+   */
+  @Fluent
+  PoolConfig connectingTo(SqlConnectOptions server);
+
+  /**
+   * Configures the pool to pick the server among the given list of {@code servers} using round robin
+   * load balancing.
+   *
+   * <ul>
+   *   <li>The first server of the list defines the base options applied to all servers and defines the common connection configuration such as SSL certificates, etc...</li>
+   *   <li>The remaining servers use the network address, user, password and databases</li>
+   * </ul>
+   * <p>
+   *
+   * @param servers the list of servers
+   * @return a reference to this, so the API can be used fluently
+   */
+  @Fluent
+  PoolConfig connectingTo(List<SqlConnectOptions> servers);
+
+  /**
+   * Configures the pool to user a dynamic configuration provided by the {@code serverProvider}.
+   *
+   * <ul>
+   *   <li>The {@code base} options is applied to all servers and defines the common connection configure such as SSL certificates, etc...</li>
+   *   <li>When the pool creates a connection it uses the value returned by the supplier to define the network address, user, password and databases to use</li>
+   * </ul>
+   * <p>
+   *
+   * When the {@code serverProvider} fails to return a value, the pool falls back to the {@code base} options.
+   *
+   * @param base the base options for all servers
+   * @param serverProvider the list of servers
+   * @return a reference to this, so the API can be used fluently
+   */
+  @GenIgnore
+  <C extends SqlConnectOptions> PoolConfig connectingTo(C base, Supplier<C> serverProvider);
```

Comment:
In our case getting connection options could involve io, is it possible to make it async friendly, i.e. provide `Supplier<Future<C>>` or something similar?

Reply:
I think that would be too complicated, instead the supplier should communicate asynchronously with something else that retrieves the options, e.g the options are set in a volatile field and the provider reads that field

Reply:
yes, it would be more complicated  
but without it user have to wrap Pool and all its methods that could create a new connection and add its async logic around that would refresh `SqlConnectOptions` and retry. (at least this would happen in my case since credentials storage does not provide notifications mechanism...)  

As alternative it could be added a hook that would be invoked before an attempt of creating a new connection. Would it be  possible? 

Anyway `Supplier<C>` is better than nothing, at least now it is possible to not recreate entire Pool.

Reply:
I will try to use a `Supplier<Future<>>` as it seems legit

Reply:
now it uses a future, can you have a look ?

Reply:
looks great, thanks